src/ngraph_backend_manager.cc

```diff
@@ -0,0 +1,62 @@
+/*******************************************************************************
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#include "ngraph_backend_manager.h"
+
+#include <iostream>
+
+namespace tensorflow {
+namespace ngraph_bridge {
+
+BackendManager* BackendManager::Instance() {
+  static BackendManager instance;
+  return &instance;
+}
+
+void BackendManager::SetBackendConfig(const std::string& config) {
+  std::lock_guard<std::mutex> lock{mu_};
+  LoadBackendLocked(config);
+}
+
+std::tuple<std::shared_ptr<ngraph::runtime::Backend>, bool>
+BackendManager::GetBackend() {
+  std::lock_guard<std::mutex> lock{mu_};
+
+  if (!backend_) {
+    const char* config = std::getenv("NGRAPH_TF_BACKEND");
+    if (!config) {
+      config = "CPU";
+    }
+    LoadBackendLocked(config);
+  }
+
+  auto colon = config_.find(':');
+  bool is_cpu;
+  if (colon == std::string::npos) {
+    is_cpu = config_ == "CPU";
+  } else {
+    is_cpu = config_.substr(0, colon) == "CPU";
+  }
+
+  return std::make_tuple(backend_, is_cpu);
+}
+
+void BackendManager::LoadBackendLocked(std::string config) {
+  backend_ = ngraph::runtime::Backend::create(config);
```

Comment:
The function is named `LoadBackendLocked` but the mutex lock is in `SetBackendConfig` and `GetBackend`. I suppose the mutex is there to guard this line `backend_ = ngraph::runtime::Backend::create(config);`. In which case, can we have the `lock_guard` in `LoadBackendLocked`

Reply:
It's there to ensure that `config_` is set atomically with `backend_`. `GetBackend` checks `backend_` before calling `SetBackendLocked`; it needs to hold the mutex, ideally without dropping it between checking `backend_` and setting `backend_`.

(The naming convention comes from the NT kernel and low-level runtime -- "*Locked" means, to the reader, "Make sure the appropriate lock is held at all calls to this function."  :-)  I'm happy to rename it if another convention would make it more clear.)

Reply:
right. I realized after typing my comment that `std::getenv("NGRAPH_TF_BACKEND");` could cause races (thread1 reads backend = CPU, the env var changes to GPU, thread2 reads GPU. thread2 sets to GPU, thread 1 resets to CPU. Even though it should have been GPU it got set to CPU) after I wrote the above comment, but forgot to remove it from the review.

Interesting new info about the `*Locked` naming convention in NT. Since we do not have any such convention going, am happy with the name.